File: examples/csharp/Customer/src/CustomerLogic.cs
using Angzarr;
using Examples;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Serilog;

namespace Angzarr.Examples.Customer;

/// <summary>
/// Default implementation of customer business logic.
/// </summary>
public class CustomerLogic : ICustomerLogic
{
    private readonly Serilog.ILogger _logger;

    public CustomerLogic(Serilog.ILogger logger)
    {
        _logger = logger.ForContext<CustomerLogic>();
    }

    public CustomerState RebuildState(EventBook? eventBook)
    {
        var state = CustomerState.Empty;

        if (eventBook == null || eventBook.Pages.Count == 0)
            return state;

        // Start from snapshot if present
        if (eventBook.Snapshot?.State != null)
        {
            if (eventBook.Snapshot.State.TryUnpack<global::Examples.CustomerState>(out var snapState))
            {
                state = new CustomerState(
                    snapState.Name,
                    snapState.Email,
                    snapState.LoyaltyPoints,
                    snapState.LifetimePoints);
            }
        }

        // Apply events
        foreach (var page in eventBook.Pages)
        {
            if (page.Event == null) continue;
            state = ApplyEvent(state, page.Event);
        }

        return state;
    }

    private CustomerState ApplyEvent(CustomerState state, Any eventAny)
    {
        var typeUrl = eventAny.TypeUrl;

        if (typeUrl.EndsWith("CustomerCreated"))
        {
            var evt = eventAny.Unpack<CustomerCreated>();
            return new CustomerState(evt.Name, evt.Email, 0, 0);
        }

        if (typeUrl.EndsWith("LoyaltyPointsAdded"))
        {
            var evt = eventAny.Unpack<LoyaltyPointsAdded>();
            return state
                .WithLoyaltyPoints(evt.NewBalance)
                .AddLifetimePoints(evt.Points);
        }

        if (typeUrl.EndsWith("LoyaltyPointsRedeemed"))
        {
            var evt = eventAny.Unpack<LoyaltyPointsRedeemed>();
            return state.WithLoyaltyPoints(evt.NewBalance);
        }

        return state;
    }

    public EventBook HandleCreateCustomer(CustomerState state, string name, string email)
    {
        if (state.Exists)
            throw CommandValidationException.FailedPrecondition("Customer already exists");

        if (string.IsNullOrWhiteSpace(name))
            throw CommandValidationException.InvalidArgument("Customer name is required");

        if (string.IsNullOrWhiteSpace(email))
            throw CommandValidationException.InvalidArgument("Customer email is required");

        _logger.Information("creating_customer {@Data}", new { name, email });

        var evt = new CustomerCreated
        {
            Name = name,
            Email = email,
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleAddLoyaltyPoints(CustomerState state, int points, string reason)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Customer does not exist");

        if (points <= 0)
            throw CommandValidationException.InvalidArgument("Points must be positive");

        var newBalance = state.LoyaltyPoints + points;

        _logger.Information("adding_loyalty_points {@Data}",
            new { points, new_balance = newBalance, reason });

        var evt = new LoyaltyPointsAdded
        {
            Points = points,
            NewBalance = newBalance,
            Reason = reason ?? ""
        };

        return CreateEventBook(evt);
    }

    public EventBook HandleRedeemLoyaltyPoints(CustomerState state, int points, string redemptionType)
    {
        if (!state.Exists)
            throw CommandValidationException.FailedPrecondition("Customer does not exist");

        if (points <= 0)
            throw CommandValidationException.InvalidArgument("Points must be positive");

        if (points > state.LoyaltyPoints)
            throw CommandValidationException.FailedPrecondition(
                $"Insufficient points: have {state.LoyaltyPoints}, need {points}");

        var newBalance = state.LoyaltyPoints - points;

        _logger.Information("redeeming_loyalty_points {@Data}",
            new { points, new_balance = newBalance, redemption_type = redemptionType });

        var evt = new LoyaltyPointsRedeemed
        {
            Points = points,
            NewBalance = newBalance,
            RedemptionType = redemptionType ?? ""
        };

        return CreateEventBook(evt);
    }

    private static EventBook CreateEventBook(IMessage evt)
    {
        var page = new EventPage
        {
            Num = 0,
            Event = Any.Pack(evt),
            CreatedAt = Timestamp.FromDateTime(DateTime.UtcNow)
        };

        var book = new EventBook();
        book.Pages.Add(page);
        return book;
    }
}

File: examples/cpp/customer/src/customer_logic.cc
#include "customer_logic.hpp"
#include "validation_error.hpp"
#include <chrono>

namespace customer {

using namespace angzarr;

CustomerState CustomerLogic::rebuild_state(const angzarr::EventBook* event_book) {
    CustomerState state;

    if (!event_book || event_book->pages().empty()) {
        return state;
    }

    // Handle snapshot if present
    if (event_book->has_snapshot() && event_book->snapshot().has_state()) {
        const auto& snap = event_book->snapshot().state();
        if (snap.type_url().find("CustomerState") != std::string::npos) {
            examples::CustomerState snap_state;
            snap.UnpackTo(&snap_state);
            state.name = snap_state.name();
            state.email = snap_state.email();
            state.loyalty_points = snap_state.loyalty_points();
            state.lifetime_points = snap_state.lifetime_points();
        }
    }

    // Apply events
    for (const auto& page : event_book->pages()) {
        if (page.has_event()) {
            state = apply_event(std::move(state), page.event());
        }
    }

    return state;
}

examples::CustomerCreated CustomerLogic::handle_create_customer(
    const CustomerState& state, const std::string& name, const std::string& email) {
    if (state.exists()) {
        throw ValidationError::failed_precondition("Customer already exists");
    }
    if (name.empty()) {
        throw ValidationError::invalid_argument("Customer name is required");
    }
    if (email.empty()) {
        throw ValidationError::invalid_argument("Customer email is required");
    }

    examples::CustomerCreated event;
    event.set_name(name);
    event.set_email(email);
    event.mutable_created_at()->set_seconds(
        std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()));
    return event;
}

examples::LoyaltyPointsAdded CustomerLogic::handle_add_loyalty_points(
    const CustomerState& state, int32_t points, const std::string& reason) {
    if (!state.exists()) {
        throw ValidationError::failed_precondition("Customer does not exist");
    }
    if (points <= 0) {
        throw ValidationError::invalid_argument("Points must be positive");
    }

    examples::LoyaltyPointsAdded event;
    event.set_points(points);
    event.set_new_balance(state.loyalty_points + points);
    event.set_reason(reason);
    return event;
}

examples::LoyaltyPointsRedeemed CustomerLogic::handle_redeem_loyalty_points(
    const CustomerState& state, int32_t points, const std::string& redemption_type) {
    if (!state.exists()) {
        throw ValidationError::failed_precondition("Customer does not exist");
    }
    if (points <= 0) {
        throw ValidationError::invalid_argument("Points must be positive");
    }
    if (points > state.loyalty_points) {
        throw ValidationError::failed_precondition(
            "Insufficient points: have " + std::to_string(state.loyalty_points) +
            ", need " + std::to_string(points));
    }

    examples::LoyaltyPointsRedeemed event;
    event.set_points(points);
    event.set_new_balance(state.loyalty_points - points);
    event.set_redemption_type(redemption_type);
    return event;
}

CustomerState CustomerLogic::apply_event(CustomerState state, const google::protobuf::Any& event) {
    const auto& type_url = event.type_url();

    if (type_url.find("CustomerCreated") != std::string::npos) {
        examples::CustomerCreated e;
        event.UnpackTo(&e);
        state.name = e.name();
        state.email = e.email();
    } else if (type_url.find("LoyaltyPointsAdded") != std::string::npos) {
        examples::LoyaltyPointsAdded e;
        event.UnpackTo(&e);
        state.loyalty_points = e.new_balance();
        state.lifetime_points += e.points();
    } else if (type_url.find("LoyaltyPointsRedeemed") != std::string::npos) {
        examples::LoyaltyPointsRedeemed e;
        event.UnpackTo(&e);
        state.loyalty_points = e.new_balance();
    }

    return state;
}

}  // namespace customer
